File: model/src/main/scala/ru/mipt/acsl/decode/model/component/Command.scala
package ru.mipt.acsl.decode.model.component

import ru.mipt.acsl.decode.model._
import ru.mipt.acsl.decode.model.naming.{Container, ElementName, HasName}
import ru.mipt.acsl.decode.model.proxy.MaybeProxy
import ru.mipt.acsl.decode.model.types.{Alias, DecodeType, TypeMeasure}

/**
  * @author Artem Shein
  */
trait Command extends Referenceable with Container with HasName with MayHaveId with HasInfo {

  def alias: Alias.ComponentCommand

  def returnTypeUnit: TypeMeasure

  def returnTypeProxy: MaybeProxy.TypeProxy = returnTypeUnit.typeProxy

  def returnType: DecodeType = returnTypeProxy.obj

  def parameters: Seq[Parameter] = objects.flatMap {
    case p: Parameter => Seq(p)
    case _ => Seq.empty
  }

  override def name: ElementName = alias.name

  override def info: LocalizedString = alias.info

}

object Command {

  private case class CommandImpl(alias: Alias.ComponentCommand, id: Option[Int],
                          var objects: Seq[Referenceable], returnTypeUnit: TypeMeasure)
    extends Command

  def apply(alias: Alias.ComponentCommand, id: Option[Int],
            objects: Seq[Referenceable], returnTypeUnit: TypeMeasure): Command =
    CommandImpl(alias, id, objects, returnTypeUnit)
}

File: model/src/main/java/ru/mipt/acsl/decode/model/component/Command.java
package ru.mipt.acsl.decode.model.component;

import org.jetbrains.annotations.Nullable;
import ru.mipt.acsl.decode.model.*;
import ru.mipt.acsl.decode.model.naming.Container;
import ru.mipt.acsl.decode.model.naming.ElementName;
import ru.mipt.acsl.decode.model.naming.HasName;
import ru.mipt.acsl.decode.model.proxy.MaybeTypeProxy;
import ru.mipt.acsl.decode.model.registry.Language;
import ru.mipt.acsl.decode.model.types.Alias;
import ru.mipt.acsl.decode.model.types.DecodeType;
import ru.mipt.acsl.decode.model.types.TypeMeasure;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * Created by metadeus on 06.06.16.
 */
public interface Command extends Container, HasName, MayHaveId, HasInfo, HasAlias {

    static Command newInstance(Alias.ComponentCommand alias, Component component, @Nullable Integer id, List<Referenceable> objects,
                               TypeMeasure returnTypeMeasure) {
        return new CommandImpl(alias, component, id, objects, returnTypeMeasure);
    }

    Component component();

    @Override
    Alias.ComponentCommand alias();

    TypeMeasure returnTypeMeasure();

    default MaybeTypeProxy returnTypeProxy() {
        return returnTypeMeasure().typeProxy();
    }

    default DecodeType returnType() {
        return returnTypeProxy().obj();
    }

    default List<Parameter> parameters() {
        return objects().stream().flatMap(p -> p instanceof Parameter ? Stream.of((Parameter) p) : Stream.empty())
                .collect(Collectors.toList());
    }

    @Override
    default ElementName name() {
        return alias().name();
    }

    @Override
    default Map<Language, String> info() {
        return alias().info();
    }

    @Override
    default <T> T accept(ContainerVisitor<T> visitor) {
        return visitor.visit(this);
    }
}